### PR TITLE
Implement CloudEvents for Runs

### LINKED ISSUE #3862: Cloud Events for Runs
### Feature request

Cloud events should be sent Runs.

It needs to be identified whether we want this to be a responsibility of the custom task controllers, the pipeline controller or a dedicated one. 

### Use case

We support cloud events for taskruns and pipelineruns, we should support them for runs too.

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Emit CloudEvents for Runs. This is achieved by:
- add a new read-only controller for Runs
- emit CloudEvents only (no k8s events) on every reconcile of a Run
- use an ephemeral cache to store sent events across reconcile runs.
  This is required because since the Runs controller only observes
  Runs, it does not have the context to know what was changed in the
  Run and though if a new event is required.

The ephemeral cache logic is largely taken from the same
functionality implemented in tektoncd/experimental/cloudevents

Fixes #3862
Depends-on #4659

Signed-off-by: Andrea Frittoli <andrea.frittoli@uk.ibm.com>

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
Tekton Pipelines can now generate CloudEvents for Runs. 
This feature is disabled by default. It can by enable by setting a sink URL and flipping the new feature flag "send-cloudevents-for-runs" to "true"
```

### CODE REVIEW

**Comment by afrittoli** [comment]
This PR include one commit from #4659 as it depends on it.
Once #4659 is merge I will remove the extra commit from here.

File: pkg/reconciler/events/cache/cacheclient.go
Comment (lbernick) [inline pkg/reconciler/events/cache/cacheclient.go:38]: there's some duplication here with the`withFakeCacheClient` func below that could probably be pulled into a helper that accepts a buffer size.

File: pkg/reconciler/run/run_test.go
Comment (lbernick) [inline pkg/reconciler/run/run_test.go:135]: this line can probably be removed from the individual test cases

File: pkg/reconciler/testing/logger.go
Comment (lbernick) [inline pkg/reconciler/testing/logger.go:33]: nit: could you please add an inline comment: `lru.New(/*bufSize=*/ 128)`

**Review by lbernick — COMMENTED** [review]
I think this would be easier to review if split up into one PR that adds the run controller and one that adds the cloud events (sorry to create more work!)

File: pkg/reconciler/events/cache/cacheclient.go
Comment (afrittoli) [reply pkg/reconciler/events/cache/cacheclient.go:38]: done

File: pkg/reconciler/run/run_test.go
Comment (afrittoli) [reply pkg/reconciler/run/run_test.go:135]: good catch, done

File: pkg/reconciler/testing/logger.go
Comment (afrittoli) [reply pkg/reconciler/testing/logger.go:33]: Oh, in fact this is not needed anymore, it's handled by injection and I forgot to remove it here

File: pkg/reconciler/run/run_test.go
Comment (afrittoli) [reply pkg/reconciler/run/run_test.go:135]: done

**Comment by afrittoli** [comment]
> I think this would be easier to review if split up into one PR that adds the run controller and one that adds the cloud events (sorry to create more work!)

Hi @lbernick - I'm not sure how you would like to split this further 🙏 
The cloud events are added in https://github.com/tektoncd/pipeline/pull/4659, but I had to include that commit in this PR because it depends on it - thanks to GitHub that does not support PRs with dependencies.
Once https://github.com/tektoncd/pipeline/pull/4659 is merged, I will rebase this and it will only include adding the controller (with the cache). 

The only new part in this PR is https://github.com/tektoncd/pipeline/pull/4663/commits/749bdbae0e8f28487a66d0693770f8cd3becc282

File: pkg/reconciler/events/cache/cacheclient.go
Comment (vdemeester) [inline pkg/reconciler/events/cache/cacheclient.go:42]: Is it required in the logs as info or was it a "debug" ? 😛

**Review by vdemeester — APPROVED** [review]
One small comment, otherwise LGTM 🐯

**Comment by lbernick** [comment]
> Hi @lbernick - I'm not sure how you would like to split this further 🙏 

I was thinking moving the contents of controller.go and most of run.go into a separate PR (to merge before adding the emitCloudEvents functionality) however I'd be OK with lgtm-ing as is once you rebase this.

**Comment by afrittoli** [comment]
> > Hi @lbernick - I'm not sure how you would like to split this further 🙏
> 
> I was thinking moving the contents of controller.go and most of run.go into a separate PR (to merge before adding the emitCloudEvents functionality) however I'd be OK with lgtm-ing as is once you rebase this.

Oh, I see. The reason I put this together in one PR is that the only thing the controller does though is to invoke `emitCloudEvents`, so if moved that in a separate PR I wouldn't be able to test much about the reconcile behaviour in the initial PR.

File: pkg/reconciler/events/cache/cacheclient.go
Comment (afrittoli) [reply pkg/reconciler/events/cache/cacheclient.go:42]: Well, since this happens once per controller start-up, I thought it would be useful to have it at info to know that the cache was initialised.  If it should up a second time it means something is very wrong!

**Comment by afrittoli** [comment]
Note that when leader election is enabled, enqueued resources will be split into buckets using an hashing function which tasks `namespace/name` as input (https://github.com/knative/pkg/blob/ce6f2877cef94d5c91bafe2ce4acba52ee22c1fe/controller/controller.go#L425-L431). This means that the same resource will be consistently re-enqueue to the same controller for every subsequent update, which implies that the local cloudevents cache works fine in the HA multi-master case too.

If leadership for a bucket is transferred to a different controller instance, the local cloudevent cache will be lost, similar to what happens in case of controller restart, hence the caveat in the docs.

**Comment by afrittoli** [comment]
> 

Thanks @lbernick - it's rebased now - only one commit in the PR :)

File: pkg/reconciler/events/cloudevent/cloud_event_controller.go
Comment (lbernick) [reply pkg/reconciler/events/cloudevent/cloud_event_controller.go:142]: would it be possible to just add a test case similar to [this one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cloudevent/cloud_event_controller_test.go#L561-L573) for sending cloud events for a run?

Comment (afrittoli) [reply pkg/reconciler/events/cloudevent/cloud_event_controller.go:142]: Test added, I hope this look ok.

Comment (afrittoli) [reply pkg/reconciler/events/cloudevent/cloud_event_controller.go:142]: Add `TestEmitCloudEvents` as well now

**Comment by afrittoli** [comment]
Uhm, interesting, it looks like a flake in an existing unit test :S

/test pull-tekton-pipeline-unit-tests

**Pull request merged**
